src/stdlib_math_diff.fypp

```diff
@@ -0,0 +1,117 @@
+#! Inspired by original code (MIT license) written in 2016 by Keurfon Luu (keurfonluu@outlook.com)
+#! https://github.com/keurfonluu/Forlab/blob/master/src/lib/forlab.f90#L2673
+
+#:include "common.fypp"
+#:set RI_KINDS_TYPES = REAL_KINDS_TYPES + INT_KINDS_TYPES
+submodule (stdlib_math) stdlib_math_diff
+
+    implicit none
+
+contains
+
+    #! `diff` computes differences of adjacent elements of an array of the ${t1}$ type.
+    
+    #:for k1, t1 in RI_KINDS_TYPES
+    pure module function diff_1_${k1}$(x, n, prepend, append) result(y)
+        ${t1}$, intent(in) :: x(:)
+        integer, intent(in), optional :: n
+        ${t1}$, intent(in), optional :: prepend(:), append(:)
+        ${t1}$, allocatable :: y(:)
+        integer :: size_prepend, size_append, size_x
+        ${t1}$, allocatable :: work(:)
+        integer :: n_, i
+
+        n_ = optval(n, 1)
+        if (n_ <= 0) then
+            y = x
+            return
+        end if
+        
+        size_prepend = 0
+        size_append = 0
+        if (present(prepend)) size_prepend = size(prepend) 
+        if (present(append)) size_append = size(append)
+        size_x = size(x)
+        
+        if (size_x + size_prepend + size_append <= n_) then
+            allocate(y(0))
+            return
+        end if
+
+        allocate(work(size_x + size_prepend + size_append))
+        if (size_prepend > 0) work(:size_prepend) = prepend
+        work(size_prepend+1:size_prepend+size_x) = x
+        if (size_append > 0) work(size_prepend+size_x+1:) = append
+        
+        do i = 1, n_
+            work(2:) = work(2:) - work(1:size(work)-1)
```

Comment:
I don't understand why there is a loop, since `i` is not involve in the expression 
```fortran
            work(2:) = work(2:) - work(1:size(work)-1)
```

As written now, it will repear `n_` the same opereation. What do I miss?

Reply:
@jvdp1 I agree it is just repeating the difference operation `n_` times. But that is the intention -- and I can't see how this would be done more cleanly without a loop?

Reply:
Ok, I undertand now the meaning of `n_`. Thank you. Fine as is.

Reply:
When I submitted `diff` for the **first** time, I took advantage of the allocatable array feature. This is correct, but I am worried that the repeated allocation of arrays is not easy to understand, and maybe **not efficient**. (see https://github.com/fortran-lang/stdlib/pull/605#discussion_r776642724)

So the **second** time I adopted @gareth-nx  scheme and set up a `work` array. In fact, in the case of a large array `x` inputted in `diff`, an extra large array `work` was created to **increase memory usage** (about space complexity).

In addition, there are some unnecessary calculations ([see code](https://github.com/fortran-lang/stdlib/blob/8cc2d66e89512d04593689c79bea9432345d3ddf/src/stdlib_math_diff.fypp#L47)), which I updated in a recent commit.
At the same time, I use `move_alloc` to try to reduce the space complexity of the second scheme ([see code](https://github.com/fortran-lang/stdlib/blob/8cc2d66e89512d04593689c79bea9432345d3ddf/src/stdlib_math_diff.fypp#L50)), and I did a test ([see gist](https://gist.github.com/zoziha/e5986c4a9b752dbe0d95da90b6a5997e)):

```text
 Test case with N=    10000000 ; reps=          10 ; size_append_prepend=           1 ; order=           1
 >>> Time ratio (t_malloc_diff / t_diff) :   1.15813529    
 Test case with N=       10000 ; reps=       10000 ; size_append_prepend=           1 ; order=           1
 >>> Time ratio (t_malloc_diff / t_diff) :   1.50465608    
 Test case with N=         100 ; reps=     1000000 ; size_append_prepend=           1 ; order=           1
 >>> Time ratio (t_malloc_diff / t_diff) :   1.51453304    
 
 Test case with N=    10000000 ; reps=          10 ; size_append_prepend=           0 ; order=           1
 >>> Time ratio (t_malloc_diff / t_diff) :   1.21079636    
 Test case with N=       10000 ; reps=       10000 ; size_append_prepend=           0 ; order=           1
 >>> Time ratio (t_malloc_diff / t_diff) :   1.12502217    
 Test case with N=         100 ; reps=     1000000 ; size_append_prepend=           0 ; order=           1
 >>> Time ratio (t_malloc_diff / t_diff) :   1.47107768    
 
 Test case with N=       10000 ; reps=       10000 ; size_append_prepend=           1 ; order=          10
 >>> Time ratio (t_malloc_diff / t_diff) :   1.05520868    
 Test case with N=       10000 ; reps=       10000 ; size_append_prepend=           1 ; order=         100
 >>> Time ratio (t_malloc_diff / t_diff) :  0.971785545    
 
 Test case with N=       10000 ; reps=       10000 ; size_append_prepend=           0 ; order=          10
 >>> Time ratio (t_malloc_diff / t_diff) :  0.970007539    
 Test case with N=       10000 ; reps=       10000 ; size_append_prepend=           0 ; order=         100
 >>> Time ratio (t_malloc_diff / t_diff) :  0.933800876    
```
Using `move_alloc`, in the demonstration of `reps=1,000,000` on the same array, it obviously took more time. This is normal, it will lower the memory usage, but increase the calculation time.